Add md4c, pymd4c and mdpo packages

EuAndreh opened the pull request:
###### Motivation for this change

Add `mdpo` package.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Execution of binary tested with:
```
nix-shell -I nixpkgs=. -p md4c --run 'md2html -h'
nix-shell -I nixpkgs=. -p 'python3.withPackages (p: [ p.mdpo ])' --run 'md2po --help'
```

EuAndreh committed:
maintainers: add euandreh

EuAndreh committed:
md4c: init at 0.4.6

EuAndreh committed:
pymd4c: init at 0.4.6.0b1

EuAndreh committed:
python3Packages.mdpo: init at 0.3.5

jonringer commented on maintainers/maintainer-list.nix:
```diff
@@ -2733,6 +2733,10 @@
   };
+  euandreh = {
+    name = "EuAndreh";
+    email = "eu@euandre.org";
```
please add github (name) and githubId

jonringer commented on pkgs/development/libraries/md4c/default.nix:
```diff
@@ -0,0 +1,23 @@
+
+stdenv.mkDerivation rec {
+  pname = "md4c";
+  version = "release-0.4.6";
```
nit
```suggestion
  version = "release-0.4.6";

```

jonringer replied:
```suggestion
  version = "0.4.6";
```
tag prefixes don't belong in version field

jonringer commented on pkgs/development/libraries/md4c/default.nix:
```diff
@@ -0,0 +1,23 @@
+  src = fetchFromGitHub {
+    owner = "mity";
+    repo = "md4c";
+    rev = version;
```
tag prefixes go here
```suggestion
    rev = "release-${version}";
```

jonringer commented on pkgs/development/libraries/md4c/default.nix:
```diff
@@ -0,0 +1,23 @@
+    repo = "md4c";
+    rev = version;
+    sha256 = "0km84rmcrczq4n87ryf3ffkfbjh4jim361pbld0z8wgp60rz08dh";
+  };
```
nit
```suggestion
  };

```

jonringer commented on pkgs/development/python-modules/mdpo/default.nix:
```diff
@@ -0,0 +1,23 @@
+
+buildPythonPackage rec {
+  pname = "mdpo";
+  version = "0.2.36";
```
```suggestion
  version = "0.2.36";

```

jonringer commented on pkgs/development/python-modules/mdpo/default.nix:
```diff
@@ -0,0 +1,23 @@
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "0gp8wi5c2x3yqslipvkm8m1wvnyji5g52v4195871dhb93kzyzv0";
+  };
```
```suggestion
  };

```

jonringer commented on pkgs/development/python-modules/mdpo/default.nix:
```diff
@@ -0,0 +1,23 @@
+, pymd4c
+}:
+
+buildPythonPackage rec {
+  pname = "mdpo";
```
To comply with [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#submitting-changes) please have the commit message name be of the format
```
<pkg-name>: <subject-line>
```
for more examples, please look at https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#submitting-changes

in your case, the commit message should be:
```
python3Packages.mdpo: init at 0.2.36
```

jonringer commented on pkgs/development/python-modules/pymd4c/default.nix:
```diff
@@ -0,0 +1,26 @@
+{ stdenv
+, pkgs
```
please avoid bringing in an entire package set
```suggestion
, pkg-config  # pkgs
, pkgconfig   # pythonPackages
```

jonringer commented on pkgs/development/python-modules/pymd4c/default.nix:
```diff
@@ -0,0 +1,26 @@
+    inherit pname version;
+    sha256 = "07s3arn85ri92im6x3ipljdmrxmpik7irs06i6lm17j1x6j9841d";
+  };
+  nativeBuildInputs = [ pkgs.pkgconfig ];
```
for new usage of pkgconfig, we prefer the name over the alias
```suggestion
  nativeBuildInputs = [ pkg-config ];
```

jonringer commented on pkgs/development/python-modules/pymd4c/default.nix:
```diff
@@ -0,0 +1,26 @@
+
+buildPythonPackage rec {
+  pname = "pymd4c";
+  version = "0.4.6.0b1";
```
nit
```suggestion
  version = "0.4.6.0b1";

```

jonringer commented on pkgs/development/python-modules/pymd4c/default.nix:
```diff
@@ -0,0 +1,26 @@
+    sha256 = "07s3arn85ri92im6x3ipljdmrxmpik7irs06i6lm17j1x6j9841d";
+  };
+  nativeBuildInputs = [ pkgs.pkgconfig ];
+  propagatedBuildInputs =
+    [ md4c python3Packages.pkgconfig ];
```
don't import from other python package sets, or else you get will python38 packages in a python37 package.
```suggestion
  propagatedBuildInputs = [ md4c pkgconfig ];
```

jonringer commented on pkgs/development/python-modules/pymd4c/default.nix:
```diff
@@ -0,0 +1,26 @@
+, pkgs
+, buildPythonPackage
+, fetchPypi
+, python3Packages
```
avoid importing from other python package sets
```suggestion
```

jonringer commented on maintainers/maintainer-list.nix:
```diff
@@ -2733,6 +2733,10 @@
       fingerprint = "67FE 98F2 8C44 CF22 1828  E12F D57E FA62 5C9A 925F";
     }];
   };
+  euandreh = {
```
commit message should be:
```
maintainers: add euandreh
```

jonringer commented on pkgs/top-level/python-packages.nix:
```diff
@@ -5183,6 +5185,8 @@ in {
 
   pymc3 = callPackage ../development/python-modules/pymc3 { };
 
+  pymd4c = callPackage ../development/python-modules/pymd4c { };
```
```suggestion
  pymd4c = callPackage ../development/python-modules/pymd4c {
    inherit (pkgs) pkg-config;
  };
```

jonringer reviewed (commented):
thanks for opening your first PR to nixpkgs.

Whats the value added to nixpkgs by adding this? the package seems to be a personal project with no release management (other than uploads to pypi). https://github.com/mondeja/mdpo

EuAndreh commented:
Thanks for the review, @jonringer.

I've addressed your comments and rebased. I've only responded to the
comment on `githubId`.


Jonathan Ringer <notifications@github.com> writes:

>> @@ -2733,6 +2733,10 @@
>        fingerprint = "67FE 98F2 8C44 CF22 1828  E12F D57E FA62 5C9A 925F";
>      }];
>    };
> +  euandreh = {
> +    name = "EuAndreh";
> +    email = "eu@euandre.org";
>
> please add github (name) and githubId
>

I didn't want to tie being a package maintainer with my GitHub profile.
The file says this is an optional step [0], so I chose to not add it.

[0]: https://github.com/nixos/nixpkgs/blob/6e0320ebc12711f1f648f25e6cc497e49beb5004/maintainers/maintainer-list.nix#L8-L10

EuAndreh commented:
Jonathan Ringer <notifications@github.com> writes:

> thanks for opening your first PR to nixpkgs.

Actually, this is my second. If you're inclined to review that too,
here's a shameless plug of it: https://github.com/NixOS/nixpkgs/pull/99544

EuAndreh commented:
Jonathan Ringer <notifications@github.com> writes:

> Whats the value added to nixpkgs by adding this? the package seems to
> be a personal project with no release management (other than uploads
> to pypi). https://github.com/mondeja/mdpo

TBH, I didn't weighted this too much when packaging it. I haven't
encountered this type of discussion before, too.

Both PRs I've done are valuable tools for doing translations, and I find
them to be useful in a general sense, rather than personally only to me.

I have on my OS a few derivations packaged just for me and a few with
tweaked attributes, but both `sphinx-intl` and `mdpo` I don't think fit
the same category.

That is the value it adds to nixpkgs: translation tools I've packaged
for myself made available for other people.

What are you looking for other than the pypi releases?

jonringer commented:
> What are you looking for other than the pypi releases?

release tags or branches. Something that says, "I want people to package this".

mondeja commented:
> release tags or branches. Something that says, "I want people to package this".

Hello, I'm the maintainer of `mdpo` package. The release process will be updated to tags soon, but I've maintained using only PyPI because the package have been in beta status, with breaking changes even without major versions updates. I can report it here when will be updated to tags.

EuAndreh commented:
Álvaro Mondéjar <notifications@github.com> writes:

>> release tags or branches. Something that says, "I want people to package this".
>
> Hello, I'm the maintainer of `mdpo` package. The release process will
> be updated to tags soon, but I've maintained using only PyPI because
> the package have been in beta status, with breaking changes even
> without major versions updates. I can report it here when will be
> updated to tags. 

Well, this definitively counts as "I want people to package this".

mondeja commented:
Done, as of [v0.3.0](https://github.com/mondeja/mdpo/releases/tag/v0.3.0), mdpo releases will be made through tags and it will strictly comply with semantic versioning.

jonringer commented:
Well, this is actually very re-assuring

jonringer commented on pkgs/development/python-modules/mdpo/default.nix:
```diff
@@ -0,0 +1,25 @@
+, pymd4c
+}:
+
+buildPythonPackage rec {
+  pname = "mdpo";
```
from the description, this sounds like it's meant to be used an cli utility.

please add the following to `pkgs/top-level/all-packages.nix` to export this as an application:
```
  mdpo = with python3Packages; toPythonApplication mdpo;
```

jonringer commented:
> TBH, I didn't weighted this too much when packaging it. I haven't encountered this type of discussion before, too.

The reason why I brought this up, is there's a costs in adding a package to nixpkgs (both in resources, and peoples time). The project should be "useful to many". This explanation should be added to the "motivation for this change" section of the PR template.

EuAndreh commented:
Jonathan Ringer <notifications@github.com> writes:

> This explanation should be added to the "motivation for this change"
> section of the PR template.

ACK. Will do next time.

I've addressed the comments and updated to 0.3.0.

EuAndreh commented:
Now `mdpo` is also accessible as a top-level package:

```shell
nix-shell -I nixpkgs=. -p mdpo --run 'md2po --help'
```

EuAndreh commented:
Also thanks @mondeja for the responsiveness!

jonringer commented:
> I didn't want to tie being a package maintainer with my GitHub profile. The file says this is an optional step [0], so I chose to not add it.

I'll try to find related topic, but they are needed as we had some issues with people changing their name and not being able to find them easily

jonringer commented on maintainers/maintainer-list.nix:
```diff
@@ -2733,6 +2733,10 @@
       fingerprint = "67FE 98F2 8C44 CF22 1828  E12F D57E FA62 5C9A 925F";
     }];
   };
+  euandreh = {
+    name = "EuAndreh";
+    email = "eu@euandre.org";
+  };
```
```suggestion
  euandreh = {
    name = "EuAndreh";
    email = "eu@euandre.org";
    githubId = 2935736;
    github = "EuAndreh";
  };
```

SuperSandro2000 replied:
@EuAndreh please take a look at jonringer's review comments.

EuAndreh commented:
EuAndreh <eu@euandre.org> writes:

>
> Well, this is exactly what I mean. I'll first delete my GitHub account
> than my email address. So the email is more important than the GitHub account.

I don't want it to block this contribution, though. If you're willing to
merge withouth the `githubId`, great. Otherwise I'll accept your
suggestion, too

EuAndreh commented:
Jonathan Ringer <notifications@github.com> writes:

> I'll try to find related topic, but they are needed as we had some
> issues with people changing their name and not being able to find them
> easily

Well, this is exactly what I mean. I'll first delete my GitHub account
than my email address. So the email is more important than the GitHub account.

SuperSandro2000 commented on pkgs/development/python-modules/mdpo/default.nix:
```diff
@@ -0,0 +1,25 @@
+
+buildPythonPackage rec {
+  pname = "mdpo";
+  version = "0.3.0";
```
At least ``pymd4c`` is python3.6 plus and fails to build on python27. Can you change that and check the other packages for similar version restrictions?

mondeja replied:
mdpo is Python3.6+ also and the current version is 0.3.5

EuAndreh commented:
Sandro <notifications@github.com> writes:

> At least ``pymd4c`` is python3.6 plus and fails to build on python27.
> Can you change that and check the other packages for similar version
> restrictions?

The change is just to use `python3Packages.buildPythonPackage` instead?

SuperSandro2000 commented:
> Sandro <notifications@github.com> writes:
> At least ``pymd4c`` is python3.6 plus and fails to build on python27. Can you change that and check the other packages for similar version restrictions?
> The change is just to use `python3Packages.buildPythonPackage` instead?

```disabled = pythonOlder "3.6";```

Otherwise it could mix and break the latest python with the version before that.

EuAndreh commented:
I've added the `pythonOlder = "3.6"` restriction, updated mdpo to 3.5
and rebased with master.

EuAndreh commented:
EuAndreh <eu@euandre.org> writes:

> I've added the `pythonOlder = "3.6"` restriction, updated mdpo to 3.5
> and rebased with master.

It stopped working for some reason, let me investigate and fix it.

SuperSandro2000 commented on pkgs/development/python-modules/mdpo/default.nix:
```diff
@@ -0,0 +1,28 @@
+  pname = "mdpo";
+  version = "0.3.5";
+  disabled = pythonOlder "3.6";
+  doCheck = false;
```
Please add a comment why they are disabled and add a pythonImportsCheck.

EuAndreh commented:
Sandro <notifications@github.com> writes:

> Please add a comment why they are disabled and add a pythonImportsCheck.

Both added and rebased with master.

SuperSandro2000 commented:
Result of `nixpkgs-review pr 102668` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>md4c</li>
    <li>mdpo (python38Packages.mdpo)</li>
    <li>python37Packages.mdpo</li>
    <li>python37Packages.pymd4c</li>
    <li>python38Packages.pymd4c</li>
  </ul>
</details>

python37Packages.pymd4c.log
142:Ran 0 tests in 0.000s

python38Packages.pymd4c.log
142:Ran 0 tests in 0.000s


Please add ``doCheck = false;``` and add a pythonImportsCheck.

SuperSandro2000 commented:
Result of `nixpkgs-review pr 102668` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>mdpo (python38Packages.mdpo)</li>
    <li>python37Packages.mdpo</li>
    <li>python37Packages.pymd4c</li>
    <li>python38Packages.pymd4c</li>
  </ul>
</details>

python38Packages.pymd4c.log
187:Ran 0 tests in 0.000s

python37Packages.pymd4c.log
187:Ran 0 tests in 0.000s

Add ``doCheck = false;`` and pythonImportsCheck.

EuAndreh commented:
On Mon, Dec 28, 2020 at 11:00:30AM -0800, Sandro wrote:

> Add ``doCheck = false;`` and pythonImportsCheck.

Done, also fixed merge conflict with master.

SuperSandro2000 commented on pkgs/os-specific/linux/nixos-rebuild/default.nix:
```diff
@@ -1,14 +1,22 @@
-{ substituteAll, runtimeShell, jq, nix, lib }:
```
I think you did some mistake when rebasing. Please drop the changes to nixos-rebuild.

EuAndreh commented:
Sandro <notifications@github.com> writes:

> Add ``doCheck = false;`` and pythonImportsCheck.

Done.

EuAndreh commented:
Sandro <notifications@github.com> writes:

> I think you did some mistake when rebasing. Please drop the changes to nixos-rebuild.

Oops, fixed. TY.

> @EuAndreh please take a look at jonringer's review comments.

I did see his comment on adding the github id, but that being an
optional field [0], I'd rather not tie my contribution with my github
profile.

[0]: https://github.com/NixOS/nixpkgs/blob/57c198236a6c08f01286cfec6a5400383b561850/maintainers/maintainer-list.nix#L8-L10

SuperSandro2000 commented:
> I'd rather not tie my contribution with my github profile.

You already did by opening this PR and if you ever rename your profile we just have a hard time finding your new name. It is not impossible and just makes the process harder.
I am totally fine if you do not want to add your real name, mail address gpg key or anything else but github name and id are already associated with this PR.

EuAndreh commented:
Sandro <notifications@github.com> writes:

>> I'd rather not tie my contribution with my github profile.
>
> You already did by opening this PR and if you ever rename your profile we just have a hard time finding your new name. It is not impossible and just makes the process harder.
> I am totally fine if you do not want to add your real name, mail address gpg key or anything else but github name and id are already associated with this PR.

No hard time if I rename or delete my profile: the email will stay the same.

EuAndreh commented:
Closing stale pull request.  Future interested parties may pick this up and merge in the future.
